### PR TITLE
Add prometheus bridge

### PR DESCRIPTION
**Pull request by dashpole**
This allows applications instrumented with the prometheus client library to use OpenTelemetry exporters (e.g. OTLP).

Limitations:

* Summary metrics are dropped by the bridge.  This could be fixed if we added the summary types to our data model (which I am personally against), or we could convert the summary to individual gauge and counter series.
* Start times for histograms and counters are set to the process start time.  This would only not be correct when using custom collectors, which is relatively uncommon.  That will be fixed when the prometheus client allows recording start times for their instruments.

Despite the limitations, I think this still has a lot of value to users.

### CODE REVIEW

**Comment by bwplotka** [comment]
Epic work! 

We were actually discussing similar with Prometheus client maintainers. Not sure what form we all prefer and what's better for community, but perhaps it would make sense to either host or at least document such option on https://github.com/prometheus/client_golang (`client_golang` maintainer here). I think it would need to be a separate Go module given Otel deps, but generally seamless extension of client_golang instrumentation to push with OTLP metrics, maintained in collaboration and available to users would be nice. cc @beorn7 @kakkoyun We are open for suggestions.

As per created timestamp, we are working on it (feedback welcome):
* https://docs.google.com/document/d/1kakDVn8aP1JerimLeazuMfy2jaB14W2kZDHMRokPcv4/edit
* https://github.com/prometheus/client_golang/pull/1313

File: bridge/prometheus/producer.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+		}
+		if m.GetTimestampMs() != 0 {
+			dp.Time = time.UnixMilli(m.GetTimestampMs())
+		}
+		otelCounter.DataPoints[i] = dp
+	}
+	return otelCounter
+}
+
+func convertHistogram(metrics []*dto.Metric, now time.Time) metricdata.Histogram[float64] {
```
Comment (bwplotka) [inline bridge/prometheus/producer.go:158]: I think this does not support Native Histograms (exponential ones): https://github.com/prometheus/client_model/blob/master/io/prometheus/client/metrics.proto#L82

That's fine, but let's provide clean error/warning and TODO (:

File: bridge/prometheus/producer.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+		Metrics: otelMetrics,
+	}}, nil
+}
+
+func convertPrometheusMetricsInto(promMetrics []*dto.MetricFamily, now time.Time) ([]metricdata.Metrics, error) {
+	var errs multierr
+	otelMetrics := make([]metricdata.Metrics, 0)
+	for _, pm := range promMetrics {
+		newMetric := metricdata.Metrics{
+			Name:        pm.GetName(),
```
Comment (fatsheep9146) [inline bridge/prometheus/producer.go:94]: Is this necessary to convert prometheus metric name to otlp metrics name?
For example convert `http_request_duration_count` to `http.request.duration.count`?

File: bridge/prometheus/config.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// limitations under the License.
+
+package prometheus // import "go.opentelemetry.io/otel/bridge/prometheus"
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// config contains options for the exporter.
+type config struct {
```
Comment (fatsheep9146) [inline bridge/prometheus/config.go:22]: small nit: I think you mean "config contains options for the producer", not exporter

File: bridge/prometheus/config.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+	if len(cfg.gatherers) == 0 {
+		cfg.gatherers = []prometheus.Gatherer{prometheus.DefaultGatherer}
+	}
+
+	return cfg
+}
+
+// Option sets exporter option values.
+type Option interface {
```
Comment (fatsheep9146) [inline bridge/prometheus/config.go:41]: same here, exporter -> producer

File: bridge/prometheus/producer.go (hunk 2 above)
Comment (dashpole) [reply bridge/prometheus/producer.go:94]: Unfortunately, it isn't feasible to do that, as otel can have underscores in names as well (e.g. `my.very_special.metric`)

**Comment by dashpole** [comment]
@open-telemetry/proto-go-approvers I was thinking maybe this makes more sense in contrib?  It isn't really a "core" part of what OTel requires from SDKs.

**Comment by MrAlias** [comment]
> @open-telemetry/proto-go-approvers I was thinking maybe this makes more sense in contrib? It isn't really a "core" part of what OTel requires from SDKs.

I'm not opposed. What are your thoughts on hosting this in a personal/3rd party repository?

**Comment by dashpole** [comment]
I'm not a huge fan of personal repos, as it makes ownership changes harder on users.  The other option would be for me to ask the prometheus maintainers to host this in a Prometheus-owned repository.  When I last spoke with them, they were OK with that.  If i'm initially maintaining the bridge, it is probably easiest for me to do so in an OTel repository, as i'm an approver.

**Comment by MrAlias** [comment]
> I'm not a huge fan of personal repos, as it makes ownership changes harder on users. The other option would be for me to ask the prometheus maintainers to host this in a Prometheus-owned repository. When I last spoke with them, they were OK with that. If i'm initially maintaining the bridge, it is probably easiest for me to do so in an OTel repository, as i'm an approver.

Yeah, contrib sounds good to me then if you plan to act as a maintainer for it :+1:

**Comment by dashpole** [comment]
Superseded by https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4227

**Pull request closed without merging**